backend/src/nodes/nodes/image_adjustment/color_levels.py

```diff
@@ -0,0 +1,125 @@
+
+import numpy as np
+
+from . import category as ImageAdjustmentCategory
+from ...node_base import NodeBase
+from ...node_factory import NodeFactory
+from ...properties.inputs import ImageInput, SliderInput, BoolInput
+from ...properties.outputs import ImageOutput
+from ...impl.image_utils import as_3d
+from ...utils.utils import get_h_w_c
+
+
+@NodeFactory.register("chainner:image:color_levels")
+class ColorLevelsNode(NodeBase):
+    """
+    Color Levels can be used to make an image lighter or darker,
+    to change contrast or to correct a predominant color cast.
+
+    This code was adapted from a Stack-Overflow answer by Iperov,
+    can found at: https://stackoverflow.com/a/60339950
+    """
+
+    def __init__(self):
+        super().__init__()
+        self.description = "Adjust color levels"
+        self.inputs = [
+            ImageInput(),
+            BoolInput("Red", default=True),
+            BoolInput("Green", default=True),
+            BoolInput("Blue", default=True),
+            BoolInput("Alpha", default=False),
+            SliderInput(
+                "In Black",
+                minimum=0,
+                maximum=255,
+                default=0,
+                precision=1,
+                controls_step=1,
+            ),
+            SliderInput(
+                "In White",
+                minimum=0,
+                maximum=255,
+                default=255,
+                precision=1,
+                controls_step=1,
+            ),
+            SliderInput(
+                "Gamma",
+                minimum=0,
+                maximum=10,
+                default=1,
+                precision=3,
+                controls_step=0.01,
+                scale="log",
+            ),
+            SliderInput(
+                "Out Black",
+                minimum=0,
+                maximum=255,
+                default=0,
+                precision=1,
+                controls_step=1,
+            ),
+            SliderInput(
+                "Out White",
+                minimum=0,
+                maximum=255,
+                default=255,
+                precision=1,
+                controls_step=1,
+            ),
+        ]
+        self.outputs = [ImageOutput(image_type="Input0")]
+        self.category = ImageAdjustmentCategory
+        self.name = "Color Levels"
+        self.icon = "MdOutlineColorLens"
+        self.sub = "Adjustments"
+
+    def run(
+        self,
+        img: np.ndarray,
+        red: bool,
+        green: bool,
+        blue: bool,
+        alpha: bool,
+        in_black: float,
+        in_white: float,
+        in_gamma: float,
+        out_black: float,
+        out_white: float,
+    ) -> np.ndarray:
+        """Adjust color levels of image"""
+
+        _, _, c = get_h_w_c(img)
+        c = min(c, 4)
+
+        if c == 1:
+            img = as_3d(img)
+            red, green, blue = True, True, True
+
+        img = (img * 255).astype("uint8")
```

Comment:
Is the conversion to 8 bit really necessary? Some people use chainner to process 16 bit images.

Reply:
Removed 8 bit conversion. It was unnecessary